### PR TITLE
[Test Only] Add bfloat16 accuracy coverage for binary arithmetic

### PR DESCRIPTION
**Pull request by KalaivaniMCW**
### PR Category

Test Only
### Summary

#53613
### Notes for reviewers

- Add sufficient accuracy test coverage for Add,Sub,Mul,Divide using a pairwise binary grid 
- The grid uses every exponent above subnormal threshold and 4 mantissa bit combination per exponent.
- Test for each op also documents the underflow and overflow conditions for the device op, depending on FPU or SFPU and the difference between device and torch result.
- Remove redundant tests related to the same
- SFPU multiplication flushes subnormal boundary value (2^{-126}) ; this behaviour also reflects in the SFPU div which flushes min-normal (2^{-126})  quotients to zero. This is documented here and in the test.
- Ex: 1.175e-38 is 2^{-126} (0x0080). |0 − min| / ULP(min) = 128 ULP. The pair is 2^{-126} × 0.99609375 (0x0080 × 0x3F7F): torch keeps min; SFPU dest packing returns +0. Torch : min, device : 0 
- Ex: a = 3.39e38 / b = 1.69e38 ; torch = 2.0 and device = 0 
- The test documents the edge/boundary conditions (current SFPU and FPU behaviour) where the device output and torch output do not align for each op.
- These documented cases should be verified and validated once again in case we test on a new hardware (other than WH/BH) or if anything changes in the respective kernels or in the backend.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:kalaivani/binary_test1)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:kalaivani/binary_test1)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:kalaivani/binary_test1)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:kalaivani/binary_test1)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:kalaivani/binary_test1)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:kalaivani/binary_test1)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:kalaivani/binary_test1)

### CODE REVIEW

File: tests/ttnn/unit_tests/operations/eltwise/test_binary_category1_bfloat16.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+    golden_function = ttnn.get_golden_function(ttnn_op)
+    golden = golden_function(input_a, input_b, device=device, fast_and_approximate_mode=fast_and_approximate_mode)
+
+    tt_result = ttnn_op(tt_a, tt_b, fast_and_approximate_mode=fast_and_approximate_mode)
+    result = ttnn.to_torch(tt_result)
+
+    # Device flushes subnormal sums to zero.
+    result = flush_to_zero(result)
+    golden = flush_to_zero(golden)
```
[the remark this answers is not in the document]
Comment (KalaivaniMCW) [reply tests/ttnn/unit_tests/operations/eltwise/test_binary_category1_bfloat16.py]: for sfpu mul, we need to include the boundary value

File: tests/ttnn/unit_tests/operations/eltwise/test_add.py [hunk 2]
```diff
@@ -56,34 +56,6 @@ def test_add_1D_tensor_and_scalar(device, scalar, size):
...
-
-    input_tensor_a = ttnn.from_torch(torch_input_tensor_a, layout=ttnn.TILE_LAYOUT, device=device)
-    input_tensor_b = ttnn.from_torch(torch_input_tensor_b, layout=ttnn.TILE_LAYOUT, device=device)
-    output = ttnn.add(input_tensor_a, input_tensor_b)
-    output = ttnn.to_torch(output)
-
-    assert_with_ulp(expected_result=torch_output_tensor, actual_result=output, ulp_threshold=1)
-
-
 @pytest.mark.parametrize("hw", [(32, 64), (1, 1), (0, 0)])
```
[the remark this answers is not in the document]
Comment (KalaivaniMCW) [reply tests/ttnn/unit_tests/operations/eltwise/test_add.py:59]: those cases are covered in test_binary_bcast

File: tests/ttnn/unit_tests/operations/eltwise/test_binary_category1_bfloat16.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+Category 1: basic_binary_arithmetic
+
+ 1. ttnn.add              - Addition
+ 2. ttnn.sub              - Subtraction
+ 3. ttnn.mul              - Multiplication
+ 4. ttnn.div              - Division
+"""
+
+
+def _pairwise_inputs(include_spl_values=False):
```
[the remark this answers is not in the document]
Comment (KalaivaniMCW) [reply tests/ttnn/unit_tests/operations/eltwise/test_binary_category1_bfloat16.py]: spl values test coverage is done separately , not part of this PR

File: tests/ttnn/unit_tests/operations/eltwise/test_mul.py [hunk 4]
```diff
@@ -100,88 +53,6 @@ def test_multiply_int32_with_scalar(device, input_a, scalar):
...
-    input_tensor_a = ttnn.from_torch(torch_input_tensor_a, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
-    input_tensor_b = ttnn.from_torch(torch_input_tensor_b, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
-
-    output = ttnn.multiply(input_tensor_a, input_tensor_b, fast_and_approximate_mode=fast_and_approximate_mode)
-    output = ttnn.to_torch(output)
-
-    assert_with_ulp(expected_result=torch_output_tensor, actual_result=output, ulp_threshold=ulp_threshold)
-
-
 def test_binary_mul_bf16_scalar(device):
```
Comment (Aswinmcw) [inline tests/ttnn/unit_tests/operations/eltwise/test_mul.py:56]: `test_multiply_float32_with_scalar_sharded` is the only place in the repo that references #14840, and the only float32 scalar-mul case in this file. It's orthogonal to bf16 accuracy on three axes at once — dtype (fp32), the scalar path, and the explicit output `memory_config` — so neither the new bf16 grid nor `test_binary_bcast` replaces it. Worth keeping it (and `test_binary_mul_div_bf16`, whose 508/17/0.1 triple was a specific reported case) rather than folding them into the removal set.

File: tests/ttnn/unit_tests/operations/eltwise/test_binary_category1_bfloat16.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    #   b =  6.620178494631905e+35    (0x7AFF,             1.9921875 × 2^118)
+    #   exact a+b is still below the midpoint between max and overflow
+    #   (halfway is max + 2^119; b is 255/256 of that step), so torch RNE
+    #   returns max (0x7F7F). FPU add overflows to +inf.
+    #   The swapped order (b+a) and both-negative pair (0xFF7F + 0xFAFF →
+    #   torch -max, device -inf) are the same case.
+    #
+    # Do not flush all infs to zero — that would hide this discrepancy.
+    # Apply mask for positions where golden is already ±max and device is ±inf.
+    if fast_and_approximate_mode:
```
Comment (Aswinmcw) [inline tests/ttnn/unit_tests/operations/eltwise/test_binary_category1_bfloat16.py:71]: All four masks in this file rewrite `result` to `golden` but never assert that the masked set is non-empty, so they are one-directional: if the device is fixed, or the grid stops producing these pairs, the mask quietly matches nothing and the divergence it documents becomes untested — while the comment above it still claims the behaviour exists. Asserting the count (`assert allowed_overflow.sum() == <n>`, and likewise for the mul/div masks) would make the documented FPU behaviour a real expectation rather than a tolerance. Relatedly, `golden_bf16.abs() < 2.0**-120` in `test_mul`/`test_div` is a binade wider than the `[2^{-126}, 2^{-121}]` band the comment describes.

File: tests/ttnn/unit_tests/operations/eltwise/test_binary_category1_bfloat16.py [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+def test_div(device, fast_and_approximate_mode, ulp_threshold):
+    """Pairwise coverage of ttnn.div over the stratified bfloat16 grid.
+    FPU: 2 ULP is accepted. ULP > 2 in the underflow band is masked
+    (e.g. 2.342e-38 / 0.06226 → 8 ULP). Device +0 vs golden finite/inf
+    (recip flush / overflow-to-zero) is rewritten, as is FPU ±max vs
+    torch ±inf when ea−eb = 128 (7.96875 / 2.342e-38).
+    """
+    input_a, input_b = _pairwise_inputs()
+    # Replace ±0 to avoid dividing by zero.
+    input_b = torch.where(input_b == 0, torch.ones_like(input_b), input_b)
```
Comment (Aswinmcw) [inline tests/ttnn/unit_tests/operations/eltwise/test_binary_category1_bfloat16.py]: This never matches: the grid contains no zeros unless `include_spl_values=True`, which nothing in this file passes. As written it reads as if division by zero is being handled here — drop it, or plumb the specials flag through so it actually does something.

File: tests/ttnn/unit_tests/operations/eltwise/test_binary_category1_bfloat16.py [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+from tests.ttnn.utils_for_testing import assert_with_ulp, ulp_distance, flush_subnormal_values_to_zero
+from tests.ttnn.unit_tests.operations.eltwise.eltwise_test_utils import (
+    generate_bfloat16_binary_grid,
+    flush_to_zero,
+    to_tt_tensor,
+)
+
+pytestmark = pytest.mark.use_module_device
+
+"""
```
Comment (Aswinmcw) [inline tests/ttnn/unit_tests/operations/eltwise/test_binary_category1_bfloat16.py:17]: The op list is out of step with the tests: `ttnn.rsub` is exercised by `test_addlike_ops` but isn't listed (nor mentioned in the PR summary), and the division test uses `ttnn.divide`, not `ttnn.div`.

File: tests/ttnn/unit_tests/operations/eltwise/test_math.py [hunk 8]
```diff
@@ -288,3 +288,18 @@ def test_recip_fixed(device, h, w, fill_value):
         torch_output_tensor = golden_function(torch_input_tensor, device=device)
         assert_with_ulp(expected_result=torch_output_tensor, actual_result=output_tensor, ulp_threshold=1)
         assert_allclose(torch_output_tensor, output_tensor, atol=1e-2, rtol=1e-2)
+
+
+@pytest.mark.parametrize("s", [3, 0.5])
+@pytest.mark.parametrize("h", [64])
+@pytest.mark.parametrize("w", [128])
+def test_sub_scalar(device, s, h, w):
```
Comment (Aswinmcw) [inline tests/ttnn/unit_tests/operations/eltwise/test_math.py]: `test_math.py` is the unary-math file, so a binary tensor-scalar subtraction is hard to find here — and with `test_sub.py` deleted this becomes the only home for scalar sub. Either keep `test_sub.py` for it or move it next to the other binary scalar cases.

**Review by Aswinmcw — COMMENTED** [review]
Minor: the description body starts with a stray `₹` before `### PR Category`.

File: tests/ttnn/unit_tests/operations/eltwise/test_binary_category1_bfloat16.py (hunk 7 above)
Comment (KalaivaniMCW) [reply tests/ttnn/unit_tests/operations/eltwise/test_binary_category1_bfloat16.py:17]: updated

File: tests/ttnn/unit_tests/operations/eltwise/test_math.py (hunk 8 above)
Comment (KalaivaniMCW) [reply tests/ttnn/unit_tests/operations/eltwise/test_math.py]: moved this to test_binary_scalar

File: tests/ttnn/unit_tests/operations/eltwise/test_binary_category1_bfloat16.py (hunk 6 above)
Comment (KalaivaniMCW) [reply tests/ttnn/unit_tests/operations/eltwise/test_binary_category1_bfloat16.py]: This was to replace any padded zeroes if tensor shape changes, got it. updated the function to include zero flexibly

File: tests/ttnn/unit_tests/operations/eltwise/test_mul.py (hunk 4 above)
Comment (KalaivaniMCW) [reply tests/ttnn/unit_tests/operations/eltwise/test_mul.py:56]: this test is no longer needed

File: tests/ttnn/unit_tests/operations/eltwise/test_binary_category1_bfloat16.py (hunk 5 above)
Comment (KalaivaniMCW) [reply tests/ttnn/unit_tests/operations/eltwise/test_binary_category1_bfloat16.py:71]: This test is to document the known behaviour of the ops at this point of time. If at anything changes with the op implementation, we will redo the accuracy test and verify the validity of all these documented cases again.

File: tests/ttnn/unit_tests/operations/eltwise/test_binary_category1_bfloat16.py [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+    # 2) Overflow-to-zero: when ea+eb ≳ 145, torch saturates to ±inf but FPU
+    #    dest packing returns +0 (sign is dropped).
+    #    Example: 2^{18} × 2^{127} = 2^{145} (0x4880 × 0x7F00) → torch +inf, FPU +0.
+    #    Treat golden ±inf vs device +0 as allowed. Do not flush all infs.
+    if fast_and_approximate_mode:
+        golden_bf16 = golden.to(torch.bfloat16)
+        result_bf16 = result.to(torch.bfloat16)
+        both_finite = torch.isfinite(golden_bf16) & torch.isfinite(result_bf16)
+        above_2_ulp = both_finite & (golden_bf16.abs() < (2.0**-120)) & (ulp_distance(golden_bf16, result_bf16) > 2)
+        result = torch.where(above_2_ulp, golden, result)
```
[the remark this answers is not in the document]
Comment (KalaivaniMCW) [reply tests/ttnn/unit_tests/operations/eltwise/test_binary_category1_bfloat16.py:129]: the worst is not 8 ULP.  The one specified in the doscstring is just an example to illustrate > 2 ULP cases

**Review by Aswinmcw — APPROVED** [review]
Previous round's findings are addressed or answered. LGTM.
